scipy/stats/_stats.pyx

```diff
@@ -0,0 +1,174 @@
+import warnings
+import numpy as np
+from numpy cimport int64_t, uint64_t
+import scipy.special as special
+from collections import namedtuple
+
+cimport numpy as np
+cimport cython
+
+# Kendall's tau will be computed directly between these types.
+# Arrays of other types will be turned into a rank array using _toranks().
+
+ctypedef fused ordered0:
+    np.int32_t
+    np.int64_t
+    np.float32_t
+    np.float64_t
+
+ctypedef fused ordered1:
+    np.int32_t
+    np.int64_t
+    np.float32_t
+    np.float64_t
+
+@cython.wraparound(False)
+@cython.boundscheck(False)
+cdef inline uint64_t pairs(uint64_t l):
+    return (l * (l - 1)) // 2
+
+@cython.wraparound(False)
+@cython.boundscheck(False)
+cdef inline double stat0(uint64_t l):
+    return l * (l - 1.) * (2*l + 5)
+
+@cython.wraparound(False)
+@cython.boundscheck(False)
+cdef inline double stat1(uint64_t l):
+    return l * (l - 1.) * (l - 2)
+
+# Turn any array into an array of ranks.
+
+@cython.wraparound(False)
+@cython.boundscheck(False)
+def _toranks(x):
+    cdef long[:] perm = np.argsort(x, kind='quicksort')
+    cdef long[:] rank = np.ndarray(len(perm), dtype=long)
+    cdef uint64_t i, j = 0
+    for i in xrange(len(x) - 1):
+        rank[perm[i]] = j
+        if x[perm[i]] != x[perm[i + 1]]:
+            j += 1
+
+    rank[perm[i + 1]] = j
+    return rank
+
+KendalltauResult = namedtuple('KendalltauResult', ('correlation', 'pvalue'))
+
+@cython.wraparound(False)
+@cython.boundscheck(False)
+def _kendalltau(ordered0[:] x, ordered1[:] y):
+
+    cdef uint64_t n = np.uint64(len(x))
+    cdef long[:] temp = np.ndarray(n, dtype=long)  # support structure used by mergesort
+
+    # this closure recursively sorts sections of perm[] by comparing
+    # elements of y[perm[]] using temp[] as support
+    # returns the number of swaps required by an equivalent bubble sort
+
+    def mergesort(uint64_t offset, uint64_t length):
```

Comment:
I suggest we use timsort. There is an MIT licensed C++ version here:
https://github.com/gfx/cpp-TimSort